File: src/engine_ts/scenes/scene_layers/simple_scene_layer.ts
import { ISprite } from "../../sprites/i_sprite";
import { SceneLayer } from "./scene_layer";

export class SimpleSceneLayer {
    public base: SceneLayer;

    constructor(posX: number = 0, posY: number = 0, spriteList: ISprite[]) {
        this.base = new SceneLayer(posX, posY, spriteList);
    }

    /**********************************************************/
    /** GETTERS & SETTERS */
    /**********************************************************/

    public getPosX(): number { return this.base.getPosX(); }
    public getPosY(): number { return this.base.getPosY(); }
    public getSpriteList(): ISprite[] { return this.base.getSpriteList(); }

    public setPosX(posX: number): void { this.base.setPosX(posX); }
    public setPosY(posY: number): void { this.base.setPosY(posY); }
    public setSpriteList(spriteList: ISprite[]): void { this.base.setSpriteList(spriteList); }

    /**********************************************************/
    /** MÉTODOS */
    /**********************************************************/

    public unload(): void {
        this.base.unload();
    }

    public moveX(): void {
        for (let i = 0; i < this.base.spriteList.length; i++) {
            this.base.spriteList[i].moveX();
        }
    }
    public moveY(): void {
        for (let i = 0; i < this.base.spriteList.length; i++) {
            this.base.spriteList[i].moveY();
        }
    }

    /**********************************************************/
    /** MÉTODOS GAMELOOP */
    /**********************************************************/

    public render(ctx: CanvasRenderingContext2D, alpha: number): void {
        ctx.save(); // salva o estado da matriz de transformação do canvas
        ctx.translate(this.base.posX, this.base.posY); // move a origem do canvas para a posição da layer

        // Renderiza os sprites nas suas posições locais (relativas ao layer)
        for (let i = 0; i < this.base.spriteList.length; i++) {
            const u = this.base.spriteList[i].render(ctx, alpha);
        }

        ctx.restore(); // Restaura a matriz para não afetar as próximas Layers que forem desenhadas
    }
}

File: src/engine_zig/scenes/scene_layers/simple_scene_layer.zig
const std = @import("std");
const SceneLayer = @import("scene_layer.zig").SceneLayer;
const ISprite = @import("../../sprites/i_sprite.zig").ISprite;

/// Camada simples de cena contendo sprites e gerenciando seus ciclos de atualização e renderização
pub const SimpleSceneLayer = struct {
    base: SceneLayer,

    /// Construtor de SimpleSceneLayer
    pub inline fn init(
        pos_x: i32,
        pos_y: i32,
        sprite_list: []ISprite,
    ) SimpleSceneLayer {
        return .{
            .base = SceneLayer.init(pos_x, pos_y, sprite_list),
        };
    }

    // ========================================================================
    // GETTERS & SETTERS (INLINE)
    // ========================================================================

    pub inline fn getPosX(self: SimpleSceneLayer) i32 {
        return self.base.getPosX();
    }

    pub inline fn getPosY(self: SimpleSceneLayer) i32 {
        return self.base.getPosY();
    }

    pub inline fn setPosX(self: *SimpleSceneLayer, pos_x: i32) void {
        self.base.setPosX(pos_x);
    }

    pub inline fn setPosY(self: *SimpleSceneLayer, pos_y: i32) void {
        self.base.setPosY(pos_y);
    }

    pub inline fn getSpriteList(self: SimpleSceneLayer) []ISprite {
        return self.base.getSpriteList();
    }

    pub inline fn setSpriteList(self: *SimpleSceneLayer, sprite_list: []ISprite) void {
        self.base.setSpriteList(sprite_list);
    }

    pub inline fn unload(self: *SimpleSceneLayer) void {
        self.base.unload();
    }

    // ========================================================================
    // MÉTODOS DE MOVIMENTAÇÃO
    // ========================================================================

    pub inline fn moveX(self: *SimpleSceneLayer) void {
        for (self.base.sprite_list) |*sprite| {
            sprite.moveX();
        }
    }

    pub inline fn moveY(self: *SimpleSceneLayer) void {
        for (self.base.sprite_list) |*sprite| {
            sprite.moveY();
        }
    }

    // ========================================================================
    // MÉTODOS GAMELOOP
    // ========================================================================

    pub inline fn render(self: *SimpleSceneLayer, alpha: f32) void {
        for (self.base.sprite_list) |*sprite| {
            sprite.render(alpha);
        }
    }
};
